test/dynamo/test_ctx_manager.py

```diff
@@ -1644,6 +1723,807 @@ def f(x):
+        add: "f32[2]" = neg + acos;  neg = acos = None
+        return (add,)
+""",
+        )
+
+    @torch._dynamo.config.patch(enable_trace_contextlib=False)
+    def test_disable_trace_contextmanager(self):
+        @contextlib.contextmanager
+        def whoo(x):
+            try:
+                yield x.cos()
+            finally:
+                pass
+
+        def g(x):
+            return x.neg() + x.acos()
+
+        def f(x):
+            y = x.sin()
+            with whoo(x) as z:
+                y += g(z)
+            y += y.tan()
+            return y
+
+        x = torch.randn(2)
+        expected = f(x)
+        eager = EagerAndRecordGraphs()
+        out = torch.compile(backend=eager, fullgraph=False)(f)(x)
+        self.assertEqual(expected, out)
+        self.assertEqual(len(eager.graphs), 3)
+        self.assertExpectedInline(
+            normalize_gm(eager.graphs[0].print_readable(False)),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_: "f32[2]"):
+        l_x_ = L_x_
+
+        y: "f32[2]" = l_x_.sin();  l_x_ = None
+        return (y,)
+""",
+        )
+
+        self.assertExpectedInline(
+            normalize_gm(eager.graphs[1].print_readable(False)),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, L_args_0_: "f32[2]"):
+        l_args_0_ = L_args_0_
+
+        cos: "f32[2]" = l_args_0_.cos();  l_args_0_ = None
+        return (cos,)
+""",
+        )
+
+        self.assertExpectedInline(
+            normalize_gm(eager.graphs[2].print_readable(False)),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_: "f32[2]"):
+        l_x_ = L_x_
+
+        neg: "f32[2]" = l_x_.neg()
+        acos: "f32[2]" = l_x_.acos();  l_x_ = None
+        add: "f32[2]" = neg + acos;  neg = acos = None
+        return (add,)
+""",
+        )
+
+    @parametrize("name", ("suppress", "stdout", "stderr"))
+    def test_contextlib_suppress(self, name):
+        counters.clear()
+        eager = EagerAndRecordGraphs()
+
+        def fn(t):
+            y = t.sin()
+            # ensure we graph break on the suppress call below
+            if name == "suppress":
+                ctx = contextlib.suppress(ValueError)
+            elif name == "stdout":
+                ctx = contextlib.redirect_stdout(sys.stderr)
+            else:
+                ctx = contextlib.redirect_stderr(sys.stdout)
+
+            with ctx:
+                y += t.cos()
+            return y.tan()
+
+        t = torch.randn(2)
+        expected = fn(t)
+        got = torch.compile(backend=eager, fullgraph=False)(fn)(t)
+        self.assertEqual(expected, got)
+        self.assertEqual(len(counters["graph_break"]), 1)
+        name = f"redirect_{name}" if name in ("stdout", "stderr") else name
+        self.assertRegex(
+            next(iter(counters["graph_break"])),
+            f"<class 'contextlib.{name}'> not supported",
+        )
+
```

Comment:
Do you have a test somewhere on what happens if we try to reconstruct a contextmanager (that is, return it from a function being compiled?)

Reply:
I've added all the tests you suggested. Everything works except for one test which I'm not sure if the behavior is correct. In the test below, should `whoo` run in eager mode as `g` is decorated with `dynamo.disable`?

```python
def test_graph_break_and_disable___enter__(self):
    @contextlib.contextmanager
    def whoo(x):
        try:
            yield x + 1
        finally:
            pass

    def fn(x):
        ctx = whoo(x)
        try:
            torch._dynamo.graph_break()

            @torch._dynamo.disable
            def g():
                return ctx.__enter__()
            y = g()
        finally:
            ctx.__exit__(None, None, None)
        return y

    x = torch.tensor([1.0])
    expected = fn(x)

    eager = EagerAndRecordGraphs()
    out = torch.compile(backend=eager, fullgraph=False)(fn)(x)
    self.assertEqual(expected, out)
    self.assertEqual(len(eager.graphs), 0)
```

Reply:
I would expect the entirety of whoo to be run in eager-mode in this case. Because we don't support eager for part of the generator and Dynamo for another part of the generator

Reply:
Me too. But dynamo is still tracing it.

Reply:
For f3: that might need to be a loud error. Because we did the `__enter__` under Dynamo, and are attempting to run the `__exit__` in eager, cpython might not know that Dynamo advanced the generator